ethstorage/scanner/scanner.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022-2023, EthStorage.
+// For license information, see https://github.com/ethstorage/es-node/blob/main/LICENSE
+
+package scanner
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	es "github.com/ethstorage/go-ethstorage/ethstorage"
+	"github.com/ethstorage/go-ethstorage/ethstorage/p2p/protocol"
+)
+
+type Scanner struct {
+	worker   *Worker
+	feed     *event.Feed
+	interval time.Duration
+	ctx      context.Context
+	cancel   context.CancelFunc
+	wg       sync.WaitGroup
+	running  bool
+	mu       sync.Mutex
+	lg       log.Logger
+}
+
+func New(
+	ctx context.Context,
+	cfg Config,
+	sm *es.StorageManager,
+	loadKvFromCache func(uint64, common.Hash) []byte,
+	l1 es.Il1Source,
+	feed *event.Feed,
+	lg log.Logger,
+) *Scanner {
+	cctx, cancel := context.WithCancel(ctx)
+	scanner := &Scanner{
+		worker:   NewWorker(sm, loadKvFromCache, l1, cfg, lg),
+		feed:     feed,
+		interval: time.Minute * time.Duration(cfg.Interval),
+		ctx:      cctx,
+		cancel:   cancel,
+		lg:       lg,
+	}
+	scanner.wg.Add(1)
+	go scanner.update()
+	return scanner
+}
+
+func (s *Scanner) update() {
+	defer s.wg.Done()
+	syncEventCh := make(chan protocol.EthStorageSyncDone)
+	sub := s.feed.Subscribe(syncEventCh)
+	defer func() {
+		sub.Unsubscribe()
+		close(syncEventCh)
+		s.lg.Debug("Scanner event subscription closed")
+	}()
+
+	for {
+		s.lg.Debug("Scanner update loop")
+		select {
+		case syncDone, ok := <-syncEventCh:
+			if !ok {
+				s.lg.Debug("syncEventCh closed, exiting update loop")
+				return
+			}
+			if syncDone.DoneType == protocol.AllShardDone {
```

Comment:
What happens if a user chooses to sync and mine only a single shard while the network consists of multiple shards?

Reply:
It only checks local shards that already synced